daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+
+  gchar *data_filepath = g_file_get_path (priv->data_file);
+  gchar *metadata_filepath =
+    g_strconcat (data_filepath, METADATA_EXTENSION, NULL);
+  g_free (data_filepath);
+
+  return metadata_filepath;
+}
+
+static gboolean
+save_metadata_file (EmerCircularFile *self,
+                    GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gchar *metadata_filepath = get_metadata_filepath (self);
+  gboolean save_succeeded =
+    g_key_file_save_to_file (priv->metadata_key_file, metadata_filepath, error);
+  g_free (metadata_filepath);
+
+  return save_succeeded;
+}
+
+static gboolean
+add_to_size (EmerCircularFile *self,
+             guint64           delta,
+             GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 new_size = priv->size + delta;
+  guint64 little_endian_size = swap_bytes_64_if_big_endian (new_size);
+  g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME, SIZE_KEY,
+                         little_endian_size);
+  gboolean save_succeeded = save_metadata_file (self, error);
+  if (save_succeeded)
+    priv->size = new_size;
+
+  return save_succeeded;
+}
+
+static gboolean
+set_metadata (EmerCircularFile *self,
+              guint64           size,
+              goffset           head,
+              GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 little_endian_size = swap_bytes_64_if_big_endian (size);
+  g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME, SIZE_KEY,
+                         little_endian_size);
+  guint64 little_endian_head = swap_bytes_64_if_big_endian (head);
+  g_key_file_set_int64 (priv->metadata_key_file, METADATA_GROUP_NAME, HEAD_KEY,
+                        little_endian_head);
+
+  if (!save_metadata_file (self, error))
+    return FALSE;
+
+  priv->size = size;
+  priv->head = head;
+  return TRUE;
+}
+
+static gboolean
+read_disk_bytes (EmerCircularFile *self,
+                 guint8           *buffer,
+                 gsize             num_bytes,
+                 gsize             max_size,
+                 GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  if (num_bytes == 0)
+    return TRUE;
+
+  GFileInputStream *file_input_stream =
+    g_file_read (priv->data_file, NULL /* GCancellable */, error);
+  if (file_input_stream == NULL)
+    return FALSE;
+
+  GInputStream *input_stream = G_INPUT_STREAM (file_input_stream);
+  gsize head_to_end = max_size - priv->head;
+  gsize bytes_head;
+  if (num_bytes <= head_to_end)
+    {
+      bytes_head = num_bytes;
+    }
+  else
+    {
+      bytes_head = head_to_end;
+      gpointer copy_to = buffer + bytes_head;
+      gsize bytes_start = num_bytes - bytes_head;
```

Comment:
See other comment I posted about this calculation getting clearer if saved before the `if` block.